daemon/containerd/image_commit.go

```diff
@@ -1,17 +1,296 @@
 package containerd
 
 import (
+	"bytes"
 	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
 	"errors"
+	"fmt"
+	"runtime"
+	"time"
 
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/diff"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/containerd/rootfs"
+	"github.com/containerd/containerd/snapshots"
 	"github.com/docker/docker/api/types/backend"
+	containerapi "github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
 )
 
+/*
+This code is based on `commit` support in nerdctl, under Apache License
+https://github.com/containerd/nerdctl/blob/master/pkg/imgutil/commit/commit.go
+with adaptations to match the Moby data model and services.
+*/
+
 // CommitImage creates a new image from a commit config.
-func (i *ImageService) CommitImage(ctx context.Context, c backend.CommitConfig) (image.ID, error) {
-	return "", errdefs.NotImplemented(errors.New("not implemented"))
+func (i *ImageService) CommitImage(ctx context.Context, cc backend.CommitConfig) (image.ID, error) {
+	container := i.containers.Get(cc.ContainerID)
+
+	desc, err := i.resolveDescriptor(ctx, container.Config.Image)
+	if err != nil {
+		return "", err
+	}
+
+	cs := i.client.ContentStore()
+
+	ocimanifest, err := images.Manifest(ctx, cs, desc, platforms.DefaultStrict())
```

Comment:
@thaJeztah In the upstream we were using `ContainerConfig.Platform` https://github.com/rumpl/moby/blob/2670c94d8298523ca90db1b097d0c4ebfc3ccc56/daemon/containerd/image_commit.go#L50, but this didn't get upstreamed in https://github.com/moby/moby/pull/44804 – related comment here https://github.com/rumpl/moby/pull/98#discussion_r990187381

Reply:
Hmm.. right, would have to try to think what makes sense here; let me at least start with a quick summary of my thoughts (in case it helps);

- `docker commit` / `docker container commit` commits a _container's_ filesystem to an image
- the container would always be running a _single_ arch, so the arch should match the arch that the container was started for
- are there any cases that we can think of where the code here would be different from `ContainerConfig.Platform` ?

>  related comment here https://github.com/rumpl/moby/pull/98#discussion_r990187381

Hm, right, so that comment was about distinction between `Config` and `HostConfig`. Generally `HostConfig` would hold the "non portable" parts of the configuration; see https://github.com/moby/moby/blob/3ba527d82af53cf752aa6ec39daac14626c41b0f/api/types/container/hostconfig.go#L375-L378

Admitted, the line between "portable" and "non-portable" is sometimes "blurry" (maybe we need a better description/definition there), but I think in this case `Platform` would be considered "non-portable" (it doesn't _immediately_ define what the container looks like; I'm able to run the same container on a different architecture if the image is multi-arch, and the corresponding architecture can be resolved). (/cc @tianon @cpuguy83 for thoughts).

Reply:
I just noticed I had a pending comment in rumpl#98 >.<
I agree that having `Platform` in HostConfig would make more sense. It seems that HostConfig is not serialized though:
https://github.com/moby/moby/blob/2670c94d8298523ca90db1b097d0c4ebfc3ccc56/container/container.go#L88

And we need to persist the platform because we need it for restoring containers after daemon restart.
Possibly this should be in `Container` struct itself, if we can't have it in HostConfig?

Reply:
I don't think there's any reason to think the commited fs would have a different platform.
Also all these platform fields are codified in the OCI image config spec anyway.
If we really need it (like down the line) it could be optionally passed down from the client.